src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1192,401 @@ where
+    type Error = QueryError;
+    async fn insert_merkle_nodes<
+        Proof: MembershipProof<E, I, T> + Send + Sync + Debug + 'static,
+        E: Element + Send + Sync + Serialize,
+        I: Index + Send + Sync + Serialize,
+        T: NodeValue + Send + Sync,
+    >(
+        &mut self,
+        name: &'static str,
+        proof: Proof,
+        traversal_path: Vec<usize>,
+        block_number: u64,
+    ) -> QueryResult<()> {
+        let pos = proof.index().clone();
+        // First node in the merkle path is always the leaf node for which membership proof is provided
+        let Some(MerkleNode::Leaf { value, elem, .. }) = proof.merkle_path().first() else {
+            return Err(QueryError::Error {
+                message: "membership proof has a missing leaf".to_owned(),
+            });
+        };
+
+        let mut leaf_commit = Vec::new();
+        // Serialize the leaf node hash value into a vector
+        value
+            .serialize_compressed(&mut leaf_commit)
+            .map_err(ParseError::Serialize)?;
+
+        // Define the LTREE path to be inserted for the leaf node.
+        // Note:
+        // The traversal path is from the leaf to the root.
+        // The LTREE path is created by appending 'R' for the Root node to this path and then reversing it.
+        // The LTREE data type is treated as a string and the hierarchy of the nodes is represented by a '.' separation.
+        // For example: A traversal path [2,0,0,1] would be represented as LTREE("R.1.0.0.2").
+        let ltree_path = LTree::from(traversal_path.iter());
+        // The created field in the table is for the leaf height i.e block number
+        let created = block_number as i64;
+        let mut txn = self.transaction().await?;
+
+        // Insert the leaf hash into a separate table and use the id referenced by the state table
+        // This avoids redundancy of hashes in the state table to save space
+        let insert_hash = txn
+            .query_one(
+                "INSERT INTO hash(value) VALUES ($1) 
+                ON CONFLICT (value) DO UPDATE SET value = excluded.value 
+                RETURNING id",
+                &[&leaf_commit],
+            )
+            .await?;
+        let hash_id: i32 = insert_hash.get(0);
+
+        // upsert the leaf node
+        txn.upsert(
+            name,
+            ["pos", "created", "hash_id", "index", "entry"],
+            ["pos", "created"],
+            [[
+                sql_param(&ltree_path),
+                sql_param(&created),
+                sql_param(&hash_id),
+                sql_param(&serde_json::to_value(&pos).map_err(ParseError::Serde)?),
+                sql_param(&serde_json::to_value(elem).map_err(ParseError::Serde)?),
+            ]],
+        )
+        .await?;
+
+        let mut traversal_path = traversal_path.into_iter();
+
+        // Insert the internal nodes using the membership proof.
+        for node in proof.merkle_path().iter() {
```

Comment:
I wonder if it is possible/worthwhile to collect the values generated in this loop and do a bulk `INSERT` at the end. I think as is we could easily end up doing like 20 or 30 separate statements per update.